File: computer3/Fall Term/Solutions/Class13_Dec_04/lvl3_berry_picking.cpp
#include <bits/stdc++.h>
using namespace std;
int N, K, ans, lmt, b[1003], rem[1003];
int main(){
    scanf("%d %d", &N, &K);
    for(int i=1; i<=N; i++) {
        scanf("%d", &b[i]); lmt = max(lmt, b[i]);
    }
    for(int i=1; i<=lmt; i++){
        int cnt = 0;
        for(int j=1; j<=N; j++) cnt += b[j]/i;
        if(cnt < K/2) break;
        if(cnt >= K){ ans = max(ans, i*K/2); continue;}
        for(int j=1; j<=N; j++) rem[j] = b[j] % i;
        sort(rem+1, rem+N+1, greater<int>());
        int tmp = (cnt - K/2)*i;
        for(int j=1; j<=K-cnt; j++) tmp += rem[j];
        ans = max(ans, tmp);
    }
    printf("%d\n", ans);
}

File: computer3/Fall Term/Solutions/Class13_Dec_04/lvl3_berry_picking.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{	
		int n = readInt(), k = readInt(), b[] = new int[n], ans = 0, mx = 0;
		for(int i=0; i<n; i++) {
			b[i] = readInt();  mx = Math.max(mx, b[i]);
		}
		for(int x=1; x<=mx; x++) {
			int count = 0; List<Integer> rem = new ArrayList<>();
			for(int i=0; i<n; i++) {
				count += b[i] / x;  rem.add(b[i] % x);
			}
			if(count < k/2) break;
			if(count >= k) ans = Math.max(ans, k/2 * x);
			else {
				int sum = (count - k/2) * x;
				Collections.sort(rem, Collections.reverseOrder());
				for(int i=0; i<k - count && i < rem.size(); i++)
					sum += rem.get(i);
				ans = Math.max(ans, sum);
			}
		}
		System.out.println(ans);
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}